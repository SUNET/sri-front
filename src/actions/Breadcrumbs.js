export const MOVE_TO_DETAILS = 'MOVE_TO_DETAILS';
export const GET_OUT_OF_DETAILS = 'GET_OUT_OF_DETAILS';

export function moveToDetails(detailsEntity) {
  return {
    type: MOVE_TO_DETAILS,
    payload: {
      detailsEntity,
    },
  };
}

export function getOutOfDetails() {
  return {
    type: GET_OUT_OF_DETAILS,
  };
}
