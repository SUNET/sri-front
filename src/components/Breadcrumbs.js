import React from 'react';
import Breadcrumb from 'react-bootstrap/Breadcrumb';
import PropTypes from 'prop-types';
import Routes from '../Routes';

function generatePathsArray(pathString, isDetailsEntity, entityDetails) {
  return pathString.split('/').map((pathElement, index, array) => {
    const mathIdCurrentDetailsInLocationPath = pathString.includes(entityDetails.id);
    const isLastPath = index === array.length - 1;
    const isActive = isLastPath;
    let pathTo = `/${pathElement}`;
    let name = Routes[pathTo];

    if (typeof Routes[pathTo] === 'undefined') {
      pathTo = array.slice(0, index + 1).join('/');
      name = isLastPath && isDetailsEntity && mathIdCurrentDetailsInLocationPath ? entityDetails.name : Routes[pathTo];
    }
    return {
      pathTo,
      name,
      isActive,
    };
  });
}

const Breadcrumbs = ({ history, isDetailsEntity, entityDetails }) => {
  const { pathname } = history.location;
  const routeSteps = generatePathsArray(pathname, isDetailsEntity, entityDetails);
  return (
    <div className="breadcrumbs">
      <Breadcrumb>
        {routeSteps.map(({ pathTo, name, isActive }, index) => {
          return (
            <Breadcrumb.Item active={isActive} key={`${index * Math.random()}`} href={pathTo}>
              {name}
            </Breadcrumb.Item>
          );
        })}
      </Breadcrumb>
    </div>
  );
};

Breadcrumbs.propsTypes = {
  history: PropTypes.object.isRequired,
  isDetailsEntity: PropTypes.bool.isRequired,
  entityDetails: PropTypes.bool.isRequired,
};

export default Breadcrumbs;
