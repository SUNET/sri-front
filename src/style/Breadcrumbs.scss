@import "../variables.scss";

.breadcrumb{
    background-color: transparent;
    .breadcrumb-item{
        &:not(:first-child){
            &:before {
                content: ">";
            }
        }
    }
}
