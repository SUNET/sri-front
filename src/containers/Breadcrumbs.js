import { connect } from 'react-redux';
import Breadcrumbs from '../components/Breadcrumbs';
import { withRouter } from 'react-router-dom';

const mapStateToProps = (state, props) => {
  return {
    isDetailsEntity: state.breadcrumbs.isDetailsEntity,
    entityDetails: state.breadcrumbs.entityDetails,
    history: props.history,
  };
};

export const BreadcrumbsContainer = connect(mapStateToProps)(Breadcrumbs);

export default withRouter(BreadcrumbsContainer);
