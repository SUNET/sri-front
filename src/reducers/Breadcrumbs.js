// eslint-disable-next-line
import * as actions from '../actions/Breadcrumbs';

const appData = {
  isDetailsEntity: false,
  entityDetails: {},
};

const appReducer = (state = appData, action) => {
  switch (action.type) {
    case actions.MOVE_TO_DETAILS:
      return {
        ...state,
        isDetailsEntity: true,
        entityDetails: { ...state.entityDetails, ...action.payload.detailsEntity },
      };
    case actions.GET_OUT_OF_DETAILS:
      return {
        ...state,
        isDetailsEntity: false,
        entityDetails: {},
      };
    default:
      return state;
  }
};

export default appReducer;
