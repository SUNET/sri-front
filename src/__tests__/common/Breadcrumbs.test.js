import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Breadcrumbs from '../../components/Breadcrumbs';

const mockMath = Object.create(global.Math);
mockMath.random = () => {
  return 1;
};
global.Math = mockMath;

let BreadcrumbsWrapper;
let BreadcrumbsComponent;
const defaultProps = {
  isDetailsEntity: false,
  entityDetails: {},
  history: {
    location: {
      pathname: '/community/contacts',
    },
  },
};

const CreateBreadcrumbsWrapper = (props) => {
  const allProps = { ...defaultProps, ...props };
  return (
    <div>
      <Breadcrumbs {...allProps} />
    </div>
  );
};

const mountComponentNameWrapper = (props) => mount(<CreateBreadcrumbsWrapper {...props} />);

beforeEach(() => {
  jest.clearAllMocks();
});
afterEach(() => {
  BreadcrumbsWrapper.unmount();
});

describe('Some Component', () => {
  beforeEach(() => {
    BreadcrumbsWrapper = mountComponentNameWrapper();
    BreadcrumbsComponent = BreadcrumbsWrapper.find('Breadcrumbs');
  });
  it('Renders correctly', () => {
    expect(toJson(BreadcrumbsWrapper)).toMatchSnapshot();
    expect(BreadcrumbsComponent.find('.breadcrumbs')).toHaveLength(1);
  });
});
describe('Elements', () => {
  beforeEach(() => {
    BreadcrumbsWrapper = mountComponentNameWrapper({
      isDetailsEntity: true,
      entityDetails: {
        id: 'id-mock',
        name: 'name-mock',
      },
      history: {
        location: {
          pathname: '/community/contacts/id-mock',
        },
      },
    });
    BreadcrumbsComponent = BreadcrumbsWrapper.find('Breadcrumbs');
  });
  it('Appears Some Text in some context', () => {
    expect(toJson(BreadcrumbsWrapper)).toMatchSnapshot();
  });
});
