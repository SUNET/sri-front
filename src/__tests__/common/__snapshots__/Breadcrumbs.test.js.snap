// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Elements Appears Some Text in some context 1`] = `
<CreateBreadcrumbsWrapper
  entityDetails={
    Object {
      "id": "id-mock",
      "name": "name-mock",
    }
  }
  history={
    Object {
      "location": Object {
        "pathname": "/community/contacts/id-mock",
      },
    }
  }
  isDetailsEntity={true}
>
  <div>
    <Breadcrumbs
      entityDetails={
        Object {
          "id": "id-mock",
          "name": "name-mock",
        }
      }
      history={
        Object {
          "location": Object {
            "pathname": "/community/contacts/id-mock",
          },
        }
      }
      isDetailsEntity={true}
    >
      <div
        className="breadcrumbs"
      >
        <Breadcrumb
          label="breadcrumb"
          listProps={Object {}}
        >
          <nav
            aria-label="breadcrumb"
          >
            <ol
              className="breadcrumb"
            >
              <BreadcrumbItem
                active={false}
                href="/"
                key="0"
                linkProps={Object {}}
              >
                <li
                  className="breadcrumb-item"
                >
                  <SafeAnchor
                    href="/"
                  >
                    <a
                      href="/"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                    >
                      Home
                    </a>
                  </SafeAnchor>
                </li>
              </BreadcrumbItem>
              <BreadcrumbItem
                active={false}
                href="/community"
                key="1"
                linkProps={Object {}}
              >
                <li
                  className="breadcrumb-item"
                >
                  <SafeAnchor
                    href="/community"
                  >
                    <a
                      href="/community"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                    >
                      Community
                    </a>
                  </SafeAnchor>
                </li>
              </BreadcrumbItem>
              <BreadcrumbItem
                active={false}
                href="/community/contacts"
                key="2"
                linkProps={Object {}}
              >
                <li
                  className="breadcrumb-item"
                >
                  <SafeAnchor
                    href="/community/contacts"
                  >
                    <a
                      href="/community/contacts"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                    >
                      Contacts
                    </a>
                  </SafeAnchor>
                </li>
              </BreadcrumbItem>
              <BreadcrumbItem
                active={true}
                href="/community/contacts/id-mock"
                key="3"
                linkProps={Object {}}
              >
                <li
                  aria-current="page"
                  className="breadcrumb-item active"
                >
                  name-mock
                </li>
              </BreadcrumbItem>
            </ol>
          </nav>
        </Breadcrumb>
      </div>
    </Breadcrumbs>
  </div>
</CreateBreadcrumbsWrapper>
`;

exports[`Some Component Renders correctly 1`] = `
<CreateBreadcrumbsWrapper>
  <div>
    <Breadcrumbs
      entityDetails={Object {}}
      history={
        Object {
          "location": Object {
            "pathname": "/community/contacts",
          },
        }
      }
      isDetailsEntity={false}
    >
      <div
        className="breadcrumbs"
      >
        <Breadcrumb
          label="breadcrumb"
          listProps={Object {}}
        >
          <nav
            aria-label="breadcrumb"
          >
            <ol
              className="breadcrumb"
            >
              <BreadcrumbItem
                active={false}
                href="/"
                key="0"
                linkProps={Object {}}
              >
                <li
                  className="breadcrumb-item"
                >
                  <SafeAnchor
                    href="/"
                  >
                    <a
                      href="/"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                    >
                      Home
                    </a>
                  </SafeAnchor>
                </li>
              </BreadcrumbItem>
              <BreadcrumbItem
                active={false}
                href="/community"
                key="1"
                linkProps={Object {}}
              >
                <li
                  className="breadcrumb-item"
                >
                  <SafeAnchor
                    href="/community"
                  >
                    <a
                      href="/community"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                    >
                      Community
                    </a>
                  </SafeAnchor>
                </li>
              </BreadcrumbItem>
              <BreadcrumbItem
                active={true}
                href="/community/contacts"
                key="2"
                linkProps={Object {}}
              >
                <li
                  aria-current="page"
                  className="breadcrumb-item active"
                >
                  Contacts
                </li>
              </BreadcrumbItem>
            </ol>
          </nav>
        </Breadcrumb>
      </div>
    </Breadcrumbs>
  </div>
</CreateBreadcrumbsWrapper>
`;
